homeassistant/components/modbus/binary_sensor.py

```diff
@@ -61,10 +111,46 @@ async def async_update(self, now: datetime | None = None) -> None:
                 return
             self._lazy_errors = self._lazy_error_count
             self._attr_available = False
-            self.async_write_ha_state()
-            return
+            self._result_bits = None
+        else:
+            self._lazy_errors = self._lazy_error_count
+            self._attr_is_on = result.bits[0] & 1
+            self._attr_available = True
+            self._result_bits = result.bits
 
-        self._lazy_errors = self._lazy_error_count
-        self._attr_is_on = result.bits[0] & 1
-        self._attr_available = True
         self.async_write_ha_state()
+        if self._coordinator:
+            await self._coordinator.async_refresh()
+
+
+class SlaveSensor(CoordinatorEntity, RestoreEntity, BinarySensorEntity):
+    """Modbus slave binary sensor."""
+
+    def __init__(
+        self, coordinator: DataUpdateCoordinator[Any], idx: int, entry: dict[str, Any]
+    ) -> None:
+        """Initialize the Modbus binary sensor."""
+        idx += 1
+        self._attr_name = f"{entry[CONF_NAME]}_{idx}"
+        self._attr_device_class = entry.get(CONF_DEVICE_CLASS)
+        self._attr_should_poll = False
+        self._attr_available = False
+        self._result_inx = int(idx / 8)
+        self._result_bit = 2 ** (idx % 8)
+        super().__init__(coordinator)
+
+    async def async_added_to_hass(self) -> None:
+        """Handle entity which will be added."""
+        await super().async_added_to_hass()
+        if state := await self.async_get_last_state():
+            self._attr_is_on = state.state == STATE_ON
+            self._attr_available = True
+            self.async_write_ha_state()
+
+    @property
+    def is_on(self) -> bool | None:
+        """Return true if sensor is on."""
+        result_bits = self.coordinator.data
+        if result_bits:
+            self._attr_is_on = result_bits[self._result_inx] & self._result_bit
+        return self._attr_is_on
```

Comment:
You can do this once when the update comes in:

```suggestion
    @callback
    def _handle_coordinator_update(self) -> None:
        """Handle updated data from the coordinator."""
        result_bits = self.coordinator.data
        if result_bits:
            self._attr_is_on = result_bits[self._result_inx] & self._result_bit

        super()._handle_coordinator_update()
```

Reply:
Thanks a lot, that was exactly the function I looked for. There is a convention that function names starting with _ is class/module internal only, we should consider renaming it.

Reply:
In Java speak I would say that this method has visibility `protected`. I would consider those to still be prefixed with `_`.